frontend/packages/dev-console/src/components/pipelineruns/logs/MultiStreamLogs.tsx

```diff
@@ -0,0 +1,149 @@
+  onDownloadAll?: () => void;
+};
+
+export const MultiStreamLogs: React.FC<MultiStreamLogsProps> = ({
+  resource,
+  taskName,
+  downloadAllLabel,
+  onDownloadAll,
+}) => {
+  const scrollPane = React.useRef<HTMLDivElement>();
+  const completedRef = React.useRef<boolean[]>([]);
+  const [renderToCount, setRenderToCount] = React.useState(0);
+  const [isFullscreen, fullscreenRef, fullscreenToggle] = useFullscreen<HTMLDivElement>();
+  const [scrollDirection, handleScrollCallback] = useScrollDirection();
+  const [autoScroll, setAutoScroll] = React.useState(true);
+  const { containers, stillFetching } = getRenderContainers(resource);
+  const dataRef = React.useRef<ContainerSpec[]>(null);
+  dataRef.current = containers;
+
+  const handleComplete = React.useCallback((containerName) => {
+    const index = dataRef.current.findIndex(({ name }) => name === containerName);
+    completedRef.current[index] = true;
+    const newRenderTo = dataRef.current.findIndex((c, i) => completedRef.current[i] !== true);
+    if (newRenderTo === -1) {
+      setRenderToCount(dataRef.current.length);
+    } else {
+      setRenderToCount(newRenderTo);
+    }
+  }, []);
+
+  React.useEffect(() => {
+    scrollPane.current.onscroll = handleScrollCallback;
+  }, [handleScrollCallback]);
+
+  React.useEffect(() => {
+    if (!scrollDirection) return;
+    if (scrollDirection === ScrollDirection.scrollingUp && autoScroll === true) {
+      setAutoScroll(false);
+    }
+    if (scrollDirection === ScrollDirection.scrolledToBottom && autoScroll === false) {
+      setAutoScroll(true);
+    }
+  }, [autoScroll, scrollDirection]);
+  const downloadLogs = () => {
+    if (!scrollPane.current) return;
+    const logString = scrollPane.current.innerText;
+    const blob = new Blob([logString], {
+      type: 'text/plain;charset=utf-8',
+    });
+    saveAs(blob, `${taskName}.log`);
+  };
+
+  const containerStatus: ContainerStatus[] = resource.status?.containerStatuses ?? [];
+  const divider = <FlexItem className="odc-multi-stream-logs__divider">|</FlexItem>;
+  return (
+    <div ref={fullscreenRef} className="odc-multi-stream-logs">
+      <Flex
+        className={classNames({
+          'odc-multi-stream-logs--fullscreen': isFullscreen,
+        })}
+      >
+        <FlexItem
+          className="odc-multi-stream-logs__button"
+          breakpointMods={[{ modifier: FlexModifiers['align-right'] }]}
+        >
+          <Button variant="link" onClick={downloadLogs} isInline>
+            <DownloadIcon className="odc-multi-stream-logs__icon" />
+            Download
+          </Button>
+        </FlexItem>
+        {divider}
+        {onDownloadAll && (
+          <>
+            <FlexItem className="odc-multi-stream-logs__button">
+              <Button variant="link" onClick={onDownloadAll} isInline>
+                <DownloadIcon className="odc-multi-stream-logs__icon" />
+                {downloadAllLabel || 'Download All'}
+              </Button>
```

Comment:
Think we need to own the error handling and make a more async button here.

```suggestion
              <Button
                variant="link"
                onClick={startDownloadAll}
                isDisabled={downloadAllStatus}
                isInline
              >
                <DownloadIcon className="odc-multi-stream-logs__icon" />
                {downloadAllLabel || 'Download All'}
                {downloadAllStatus && <LoadingInline />}
              </Button>
```

```
  const [downloadAllStatus, setDownloadAllStatus] = React.useState(false);
  const startDownloadAll = () => {
    setDownloadAllStatus(true);
    onDownloadAll()
      .then(() => {
        setDownloadAllStatus(false);
      })
      .catch((err: Error) => {
        setDownloadAllStatus(false);
        const error = err.message || 'Unknown error occurred while generating the logs.';
        errorModal({ error });
      });
  };
```